### ISSUE TITLE
Lua API:  Allow to store arbitrary values in objects (meta)

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#3835]
I suggest the Lua API to be extended. Objects and especially player objects should have the ability to store arbitrary values with them and those values are to be saved in the world directory (as usual), so they are automatically preserved on disconnect, server shutdown, etc.

I think this arbitrary object data could be just an additional field called “`data`” which can take any value (preferably a table).
Or you add a setter and getter function for this additional field. Use whatever seems more reasonable.

Possible use cases:
- Store custom player attributes like mana, experience points, and many more
- Keep track of temporary effects for players, like poisoning, speed boost, etc.
- All sorts of attributes for mobs and other more sophisticated objects
- Easier to customize objects
- More clean mod code overall (no more hacky workarounds)
- Probably many more!

IMO this feature is desperately needed, because I don't know how many times I created local tables just to contain a list of players, and I even created my own files just to persist those player attributes. And it was a pain in the ass to set this all up and get (more or less) bug-free.

Also, I have seen other mods using a very creative (but also very ugly) hack in that they just abused the inventory to “store” some “data” (in form of, well, items). So people are using all sorts of workarounds already to store arbitrary data, I think this is proof enough that Minetest needs this.

What do you think? Do you approve of this suggestion?

**Comment by stujones11** [comment]
+1 While this would be a nice addition, I think what is really needed is a more generalised persistent data storage system for mods, not just limited to players/objects.

**Comment by est31** [comment]
We have the staticdata, this works already well for non-player entities. The advantage of storing it as string is of course that lua itself can decide whether to use it for a table, or for something more simple.
The disadvantage is that there is copying involved: copying into the lua VM, and then copying it out of the VM again when passed to the serialisation methods.

**Comment by stujones11** [comment]
Staticdata is not a very reliable, besides the fact that entities are still rather buggy, all data is lost when someone issues the /clearobjects command.

**Comment by Wuzzy2** [comment]
est31, staticdata does not work for players as far I know. My main concern are player objects.

**Comment by red-001** [comment]
So player metadata?

**Comment by o-jasper** [comment]
Someone came with an XP mod [in r/minetest](https://www.reddit.com/r/Minetest/comments/4gqtt6/mod_experience_expierience/), looks like they could use it.

It is preferable it would be simple. Perhaps any object you can attach data to could just have a `:static()` method, [NodeMetaRef](http://dev.minetest.net/NodeMetaRef)-equivalent functions. That seems conceptually clear?

**Comment by sofar** [comment]
https://github.com/minetest-technic/datastorage

**Comment by o-jasper** [comment]
Datastorages' approach is a nice API, but doesn't provide much structure, which depends entirely on convention. It'd be nice to automatically have some structure if the data is about a node, object, item, or player. So `:static()` or some such producing a table that is loaded when first looked at saved at the end.(could do empty⇔no file) Maybe the global `minetest.static(name)` would have a file for each name.

**Comment by sofar** [comment]
datastorage claims that it's purely for player-related data. As such it has exactly the organization that it needs: After the player name, there is no organization enforcement.

**Comment by beyondlimits** [comment]
http://irc.minetest.ru/minetest/2015-06-09#i_4283750
http://irc.minetest.ru/minetest/2016-06-20#i_4635356
http://irc.minetest.ru/minetest/2015-10-23#i_4429635
http://irc.minetest.ru/minetest/2016-11-08#i_4739365

Today I learned Lua supports weak-keyed tables. You can use objects as a keys:
```
local data = setmetatable({}, {__mode = 'k'})
data[entity] = {'key' = 'value'}
```
I'm not sure however how it would work with core internals and I have no clue how to serialize such table.

http://lua-users.org/wiki/MetatableEvents

**Comment by sofar** [comment]
@beyondlimits just use `minetest.write_json()` ?

https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L2612

**Comment by beyondlimits** [comment]
@sofar Did you mean `minetest.write_json(data)` because I don't think it would work. The keys in `data` are object, not strings or numbers.

**Comment by rubenwardy** [comment]
Since player attributes, we have meta for players - however it's worth deprecating player attributes however, and using the unified meta API for player data (and object data)

**Comment by Wuzzy2** [comment]
Closing this because 0.4.16 supports player attributes, this was all I asked for. :-)
And yes, I already made use of it, this greatly simplifies my Lua code

Other objects have `staticdata` and this is good enough for my taste. If anyone has a problem with that, probably just open a new issue as the original issue here is clearly solved now.

I don't really care about the implementation details behind this new feature, so it doesn't make any sense to keep this issue open anymore.
